alibi_detect/utils/pytorch/kernels.py

```diff
@@ -20,21 +48,122 @@ def sigma_median(x: torch.Tensor, y: torch.Tensor, dist: torch.Tensor) -> torch.
 
     Returns
     -------
-    The computed bandwidth, `sigma`.
+    The logrithm of the computed bandwidth, `log-sigma`.
     """
     n = min(x.shape[0], y.shape[0])
     n = n if (x[:n] == y[:n]).all() and x.shape == y.shape else 0
     n_median = n + (np.prod(dist.shape) - n) // 2 - 1
     sigma = (.5 * dist.flatten().sort().values[n_median].unsqueeze(dim=-1)) ** .5
-    return sigma
+    return sigma.log()
 
 
-class GaussianRBF(nn.Module):
+class KernelParameter(object):
+    """
+    Parameter class for kernels.
+    """
+
     def __init__(
         self,
-        sigma: Optional[torch.Tensor] = None,
-        init_sigma_fn: Callable = sigma_median,
-        trainable: bool = False
+        value: torch.Tensor = None,
+        init_fn: Optional[Callable] = None,
+        requires_grad: bool = False,
+        requires_init: bool = False
+    ) -> None:
+        super().__init__()
+        self.value = nn.Parameter(value if value is not None else torch.ones(1),
+                                  requires_grad=requires_grad)
+        self.init_fn = init_fn
+        self.requires_init = requires_init
+
+
+class BaseKernel(nn.Module):
+    """
+    The base class for all kernels.
+    """
+    def __init__(self) -> None:
+        super().__init__()
+        self.parameter_dict: dict = {}
+
+    def forward(self, x: Union[np.ndarray, torch.Tensor], y: Union[np.ndarray, torch.Tensor],
+                infer_parameter: bool = False) -> torch.Tensor:
+        raise NotImplementedError
+
+
+class DimensionSelectKernel(nn.Module):
+    """
+    Select a subset of the feature diomensions before apply a given kernel.
+    """
+    def __init__(self, kernel: BaseKernel, active_dims: list, feature_axis: int = -1) -> None:
+        super().__init__()
+        self.kernel = kernel
+        self.active_dims = torch.as_tensor(active_dims)
+        self.feature_axis = feature_axis
+
+    def forward(self, x: Union[np.ndarray, torch.Tensor], y: Union[np.ndarray, torch.Tensor],
+                infer_parameter: bool = False) -> torch.Tensor:
+        x, y = torch.as_tensor(x), torch.as_tensor(y)
+        if self.active_dims is not None:
+            x = torch.index_select(x, self.feature_axis, self.active_dims)
+            y = torch.index_select(y, self.feature_axis, self.active_dims)
+        return self.kernel(x, y, infer_parameter)
+
+
+class AveragedKernel(nn.Module):
+    """
+    Construct a kernel by averaging two kernels.
+
+    Parameters:
+    ----------
+        kernel_a
+            the first kernel to be averaged.
+        kernel_b
+            the second kernel to be averaged.
```

Comment:
General remark - there should be no indentation in the numpy docstring format. If you're using PyCharm you can configure the docstring format so that it is automatically generated in the correct format when an empty dosctring is first created.